pkg/tagset/slice_builder.go

```diff
@@ -0,0 +1,192 @@
+	}
+
+	for i := 0; i < levels+1; i++ {
+		bldr.offsets[i] = 0
+	}
+
+	// ensure at least the requested capacity for bldr.tags
+	if bldr.tags == nil || cap(bldr.tags) < capacity {
+		bldr.tags = make([]string, 0, capacity)
+	} else {
+		bldr.tags = bldr.tags[:0]
+	}
+
+	// ensure at least the requested capacity for bldr.hashes
+	if bldr.hashes == nil || cap(bldr.hashes) < capacity {
+		bldr.hashes = make([]uint64, 0, capacity)
+	} else {
+		bldr.hashes = bldr.hashes[:0]
+	}
+
+	// expand bldr.offsets to at least length levels
+	if bldr.levelHashes == nil || cap(bldr.levelHashes) < levels {
+		bldr.levelHashes = make([]uint64, levels)
+	} else {
+		bldr.levelHashes = bldr.levelHashes[:levels]
+	}
+
+	for i := 0; i < levels; i++ {
+		bldr.levelHashes[i] = 0
+	}
+
+	bldr.seen = map[uint64]struct{}{}
+}
+
+// Add adds the given tag to the builder at the given level.  If the tag is
+// already in the builder, it is not added again (regardless of the existing
+// tag's level).
+func (bldr *SliceBuilder) Add(level int, tag string) {
+	h := murmur3.StringSum64(tag)
+	if _, seen := bldr.seen[h]; seen {
+		return
+	}
+	bldr.seen[h] = struct{}{}
+
+	newLen := len(bldr.tags) + 1
+	insertAt := bldr.offsets[level+1]
+
+	var newTags []string
+	var newHashes []uint64
+
+	// reallocate the storage if there is not room in the existing array, and
+	// copy the existing data into the new array
+	if cap(bldr.tags) < newLen {
+		newTags = make([]string, newLen, newLen*2)
+		newHashes = make([]uint64, newLen, newLen*2)
+
+		copy(newTags[:insertAt], bldr.tags[:insertAt])
+		copy(newTags[insertAt+1:], bldr.tags[insertAt:])
+
+		copy(newHashes[:insertAt], bldr.hashes[:insertAt])
+		copy(newHashes[insertAt+1:], bldr.hashes[insertAt:])
+	} else {
+		newTags = bldr.tags[:newLen]
+		newHashes = bldr.hashes[:newLen]
+
+		// make space for the new element with the minimum number of copies,
+		// moving the first element of each level's slice to the end of that
+		// slice
+		for l := bldr.levels - 1; l > level; l-- {
+			s := bldr.offsets[l]
+			d := bldr.offsets[l+1]
+			newTags[d], newHashes[d] = newTags[s], newHashes[s]
+		}
+	}
+
+	newTags[insertAt] = tag
+	newHashes[insertAt] = h
+
+	bldr.tags = newTags
+	bldr.hashes = newHashes
+
+	bldr.levelHashes[level] ^= h
+	for l := level + 1; l < bldr.levels+1; l++ {
+		bldr.offsets[l]++
+	}
+}
+
+// AddKV adds the tag "k:v" to the builder at the given level.  If the tag is
+// already in the builder, it is not added again (regardless of the existing
+// tag's level).
+func (bldr *SliceBuilder) AddKV(level int, k, v string) {
+	tag := k + ":" + v
+	bldr.Add(level, tag)
+}
+
+// FreezeSlice "freezes" the builder and returns the requested slice of levels.
+// The Add methods cannot be called after freezing.
+func (bldr *SliceBuilder) FreezeSlice(a, b int) *Tags {
+	// free unnecessary memory, and as a convenient side-effect ensure a panic
+	// if any of the Add methods are used after this point
```

Comment:
It introduces the possibilities of "runtime" panic (e.g. after 2h of a running Agent) which I would not call convenient.

However I get the idea that once `Freeze`/`FreezeSlice` has been called, `Add*` should error, but, a panic in this situation is dangerous IMO.

Reply:
(discussed via voice) This is an example of a runtime panic from a programming error.  Ideally this panic would happen every time, such that any amount of test coverage of the caller's code would trigger the panic.  Similar things occur in the stdlib, such as `net/http`:
```
	// read fields of the request in a separate goroutine. Callers
	// should not mutate or reuse the request until the Response's
	// Body has been closed.
```

Requirements like this rely on the fallible programmer (and reviewer), and are often the source of bugs, but it's the tool we have.  What I can do to make this easier is exhort the caller to use builders start-to-finish in a single method, and provide an example of that.

Reply:
I don't think this is an issue, so there's no need to add more API surface for it.  I can remove the comment about the panic, and just include the part about saving memory.